feature/src/adm.c

```diff
@@ -48,6 +53,228 @@ static char *init_dwt_band(adm_dwt_band_t *band, char *data_top, size_t buf_sz_o
     return data_top;
 }
 
+#if ADM_OPT_ENABLE
+static char *init_dwt_band_hvd(adm_dwt_band_t *band, char *data_top, size_t buf_sz_one)
+{
+	band->band_a = NULL;
+	band->band_h = (float *)data_top; data_top += buf_sz_one;
+	band->band_v = (float *)data_top; data_top += buf_sz_one;
+	band->band_d = (float *)data_top; data_top += buf_sz_one;
+	return data_top;
+}
+
+int compute_adm(const float *ref, const float *dis, int w, int h, int ref_stride, int dis_stride, double *score, double *score_num, double *score_den, double *scores, double border_factor)
+{
+#ifdef ADM_OPT_SINGLE_PRECISION
+	double numden_limit = 1e-2 * (w * h) / (1920.0 * 1080.0);
+#else
+	double numden_limit = 1e-10 * (w * h) / (1920.0 * 1080.0);
+#endif
+	float *data_buf = 0;
+	char *data_top;
+
+	char *ind_buf_y = 0, *buf_y_orig = 0;
+	char *ind_buf_x = 0, *buf_x_orig = 0;
+	int *ind_y[4], *ind_x[4];
+
+	float *ref_scale;
+	float *dis_scale;
+
+	adm_dwt_band_t ref_dwt2;
+	adm_dwt_band_t dis_dwt2;
+
+	adm_dwt_band_t decouple_r;
+	adm_dwt_band_t decouple_a;
+
+	adm_dwt_band_t csf_r;
+	adm_dwt_band_t csf_a;
+	const float *curr_ref_scale = ref;
+	const float *curr_dis_scale = dis;
+	int curr_ref_stride = ref_stride;
+	int curr_dis_stride = dis_stride;
+
+	int orig_h = h;
+
+	int buf_stride = ALIGN_CEIL(((w + 1) / 2) * sizeof(float));
+	size_t buf_sz_one = (size_t)buf_stride * ((h + 1) / 2);
+
+	int ind_size_y = ALIGN_CEIL(((h + 1) / 2) * sizeof(int));
+	int ind_size_x = ALIGN_CEIL(((w + 1) / 2) * sizeof(int));
+
+	double num = 0;
+	double den = 0;
+
+	int scale;
+	int ret = 1;
+	
+	// Code optimized to save on multiple buffer copies 
+	// hence the reduction in the number of buffers required from 35 to 17 
+#define NUM_BUFS_ADM 17 
+	if (SIZE_MAX / buf_sz_one < NUM_BUFS_ADM)
+	{
+		printf("error: SIZE_MAX / buf_sz_one < NUM_BUFS_ADM, buf_sz_one = %zu.\n", buf_sz_one);
```

Comment:
`fprintf(stderr, ...)` ?

Reply:
This goes for all `printf()'`s. We may want to use this library with piped video data in the future, and we don't log messages mixed in.